File: implementations/fastapi/examples/src/payment/application/event/payment_cancelled_event_handler.py
from __future__ import annotations

import logging

from ....outbox.outbox_writer import OutboxWriter
from ..integration_event.payment_cancelled_integration_event import PaymentCancelledIntegrationEventV1

logger = logging.getLogger(__name__)


class PaymentCancelledEventHandler:
    """Receives the internal Domain Event (PaymentCancelled), converts it into an
    Integration Event for external BCs (payment.cancelled.v1), and loads it into the
    Outbox. The Account BC subscribes to this and executes a compensating credit (deposit)
    — a compensating transaction that reverses the amount already debited.
    """

    def __init__(self, outbox_writer: OutboxWriter) -> None:
        self._outbox_writer = outbox_writer

    async def handle(self, payload: dict) -> None:
        logger.info(
            "Payment cancelled: payment_id=%s account_id=%s reason=%s",
            payload["payment_id"],
            payload["account_id"],
            payload["reason"],
        )
        await self._outbox_writer.save_all(
            [
                PaymentCancelledIntegrationEventV1(
                    payment_id=payload["payment_id"],
                    account_id=payload["account_id"],
                    amount=payload["amount"],
                    cancelled_at=payload["cancelled_at"],
                )
            ]
        )

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/application/event/PaymentCancelledEventHandler.kt
package com.example.accountservice.payment.application.event

import com.example.accountservice.outbox.OutboxWriter
import com.example.accountservice.payment.application.integrationevent.PaymentCancelledIntegrationEventV1
import com.example.accountservice.payment.domain.PaymentCancelledEvent
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

/**
 * Receives the internal Domain Event (PaymentCancelledEvent), converts it into the Integration Event
 * for external BCs (payment.cancelled.v1), and loads it into the Outbox. Account BC subscribes to this
 * and executes the compensating credit (deposit) — a compensating transaction that reverses the
 * already-deducted amount.
 */
@Component
class PaymentCancelledEventHandler(
    private val outboxWriter: OutboxWriter,
) {
    private val logger = LoggerFactory.getLogger(PaymentCancelledEventHandler::class.java)

    fun handle(event: PaymentCancelledEvent) {
        logger
            .atInfo()
            .addKeyValue("payment_id", event.paymentId)
            .addKeyValue("account_id", event.accountId)
            .addKeyValue("reason", event.reason)
            .log("Payment cancelled")

        outboxWriter.saveAll(
            listOf(
                PaymentCancelledIntegrationEventV1(
                    paymentId = event.paymentId,
                    accountId = event.accountId,
                    amount = event.amount,
                    cancelledAt = event.cancelledAt.toString(),
                ),
            ),
        )
    }
}
